numba/cpython/builtins.py

```diff
@@ -672,3 +673,53 @@ def impl(func, iterable):
                 if func(x):
                     yield x
     return impl
+
+
+@overload(isinstance)
+def ol_isinstance(var, typs):
+    def true_impl(var, typs):
+        return True
+    def false_impl(var, typs):
+        return False
+
+    # register container types (List, Dict, ...) with types.Any as base type
+    as_numba_type.register(tuple, types.Tuple((types.Any,)))
+    as_numba_type.register(list, types.List(types.Any))
+    # dict doesn't work as input to an njit function
+    # as_numba_type.register(dict, types.DictType(types.Any, types.Any))
+
+    if isinstance(typs, types.UniTuple):
+        # corner case - all types in isinstance are the same
+        typs = (typs.key[0])
+
+    if not isinstance(typs, types.containers.Tuple):
+        typs = (typs, )
+
+    var = as_numba_type(var)
+
+    for typ in typs:
+        if isinstance(typ, types.Function):
+            key = typ.key[0]  # functions like int(..), float(..), str(..)
+        elif isinstance(typ, types.ClassType):
+            key = typ
+        else:
+            key = typ.key
+
+        numba_typ = as_numba_type(key)
+
+        if var == numba_typ:
+            return true_impl
+        elif isinstance(numba_typ, types.ClassType) and \
+                isinstance(var, types.ClassInstanceType) and \
+                var.key == numba_typ.instance_type.key:
+            # check for jitclasses
+            return true_impl
+        elif isinstance(numba_typ, types.Container) and \
+                numba_typ.key[0] == types.Any:
+            # check for containers (list, tuple, dict, ...)
+            if isinstance(var, numba_typ.__class__) or \
+                (isinstance(var, types.BaseTuple) and \
+                    isinstance(numba_typ, types.BaseTuple)):
+                return true_impl
+
+    return false_impl
```

Comment:
@stuartarchibald, I've pushed a fix to the type inference step. But I think in both cases you posted, Numba shouldn't compile the code since it cannot determine at compile-time the type of `z`.

```python
@njit
def foo(x):
    if x > 4:  # runtime information
        z = 10
    else:
        z = 'a'
    ....
```